src/treelearner/feature_histogram.hpp

```diff
@@ -80,44 +85,61 @@ class FeatureHistogram {
     }
   }
 
-  void FindBestThreshold(double sum_gradient, double sum_hessian, data_size_t num_data, double min_constraint, double max_constraint,
-                         SplitInfo* output) {
+  void FindBestThreshold(double sum_gradient, double sum_hessian,
+                         data_size_t num_data, SplitInfo *output,
+                         SplittingConstraints &constraints) {
     output->default_left = true;
     output->gain = kMinScore;
-    find_best_threshold_fun_(sum_gradient, sum_hessian + 2 * kEpsilon, num_data, min_constraint, max_constraint, output);
+    find_best_threshold_fun_(sum_gradient, sum_hessian + 2 * kEpsilon, num_data,
+                             output, constraints);
     output->gain *= meta_->penalty;
   }
 
-  void FindBestThresholdNumerical(double sum_gradient, double sum_hessian, data_size_t num_data, double min_constraint, double max_constraint,
-                                  SplitInfo* output) {
+  void FindBestThresholdNumerical(double sum_gradient, double sum_hessian,
+                                  data_size_t num_data, SplitInfo *output,
+                                  SplittingConstraints &constraints) {
     is_splittable_ = false;
+    could_be_splittable_ = false;
```

Comment:
what differences between `is_splittable_` and `could_be_splittable_ ` ?

Reply:
In the fast method (which the only thing we are looking at right now because I removed the more complicated Slow method for the moment), could_be_splittable is not used. This can be confirmed by looking at the return value of the is_splittable() function.

In the Slow method, if we decide to add it and push it, could_be_splittable is used because sometimes a split in the tree can unconstrain a leaf. This way, a leaf that was not splittable before can become splittable. Because of this, in the Slow method, we do not want to exclude for good some leaves that are not splittable right now, because they may become splittable later when the tree keeps getting built. 

Is that clear? Thanks

Reply:
okay, I see. Could you move `could_be_splittable_`  into "constraints" object? It will relief the confuse in `is_splittable_ ` and `could_be_splittable_`.